### PR TITLE
Lz4FrameEncoder should prefer direct buffers for its output.

### PR DESCRIPTION
**Pull request by normanmaurer**
Motivation:

We should prefer direct buffers for the output of Lz4FrameEncoder as this is what is needed for writing to the socket.

Modification:

Use direct buffers for the output

Result:

Less memory copies needed.

### CODE REVIEW

File: codec/src/main/java/io/netty/handler/codec/compression/Lz4FrameEncoder.java [hunk 1]
```diff
@@ -233,33 +237,25 @@ private ChannelFuture finishEncode(final ChannelHandlerContext ctx, ChannelPromi
...
+        footer.setIntLE(idx + COMPRESSED_LENGTH_OFFSET, 0);
+        footer.setIntLE(idx + DECOMPRESSED_LENGTH_OFFSET, 0);
+        footer.setIntLE(idx + CHECKSUM_OFFSET, 0);
+
         footer.writerIndex(idx + HEADER_LENGTH);
 
         compressor = null;
         checksum = null;
-        buffer = null;
+        if (buffer != null) {
```
Comment (Scottmitch) [inline codec/src/main/java/io/netty/handler/codec/compression/Lz4FrameEncoder.java]: should this be done in a finally block, and should it be created in handlerAdded and released (if not already released) in handlerRemoved?

File: codec/src/main/java/io/netty/handler/codec/compression/Lz4FrameEncoder.java [hunk 2]
```diff
@@ -233,33 +237,25 @@ private ChannelFuture finishEncode(final ChannelHandlerContext ctx, ChannelPromi
...
         final int idx = footer.writerIndex();
-        final byte[] dest = footer.array();
-        final int destOff = footer.arrayOffset() + idx;
         footer.setLong(idx, MAGIC_NUMBER);
-        dest[destOff + TOKEN_OFFSET] = (byte) (BLOCK_TYPE_NON_COMPRESSED | compressionLevel);
-        writeIntLE(0, dest, destOff + COMPRESSED_LENGTH_OFFSET);
-        writeIntLE(0, dest, destOff + DECOMPRESSED_LENGTH_OFFSET);
-        writeIntLE(0, dest, destOff + CHECKSUM_OFFSET);
+        footer.setByte(idx + TOKEN_OFFSET, (byte) (BLOCK_TYPE_NON_COMPRESSED | compressionLevel));
+        footer.setIntLE(idx + COMPRESSED_LENGTH_OFFSET, 0);
```
Comment (Scottmitch) [inline codec/src/main/java/io/netty/handler/codec/compression/Lz4FrameEncoder.java]: 0 Little Endian == 0 Big Endian ... should we just remove the `LE` and potentially avoid byte swapping?

Comment (normanmaurer) [inline codec/src/main/java/io/netty/handler/codec/compression/Lz4FrameEncoder.java]: yep let me do this...

File: codec/src/main/java/io/netty/handler/codec/compression/Lz4FrameEncoder.java (hunk 1 above)
Comment (normanmaurer) [inline codec/src/main/java/io/netty/handler/codec/compression/Lz4FrameEncoder.java]: done

**Comment by normanmaurer** [comment]
Cherry-picked into 4.1 (b42a9db8126e1caed4a1e54262fcf23413eeb82b)

**Pull request closed without merging**